github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6116381?buildTab=log) with [artifacts](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6116381?buildTab=artifacts#meta) on refs/heads/master @ [4e1bc304b29b](https://github.com/cockroachdb/pebble/commits/4e1bc304b29b):


```
db.Restart()
iter48 = db.NewIter("", "", 2 /* key types */, 0, 0, "@5" /* masking suffix */)
db.Get("mwanwaok@6")
iter49 = db.NewIter("", "", 2 /* key types */, 0, 0, "" /* masking suffix */)
iter49.Last()
iter48.Last()
iter49.SeekGE("ytzcredvf@10", "")
iter49.Next("xyjkfv@3")
db.Get("sxqbfors@1")
db.Set("mdbj@7", "ru")
iter48.Next("ohgo@2")
db.DeleteRange("jnjcss@12", "zvwggxybj@3")
db.Set("qpfthbrglrh@12", "wzpftzysohwyefbwmi")
iter48.SeekLT("eounocge@12", "")
db.Set("rigrzwjlsn@8", "ommfyt")
iter48.SeekLT("ueavly@1", "")
iter48.SeekLT("lhjdjxzgmxk@1", "axkx@8")
iter49.Prev("")
iter48.Last()
db.SingleDelete("mgtawpglvjg@4", false /* maybeReplaceDelete */)
db.Delete("wvxora@11")
iter48.SeekLT("osraunwzo@11", "")
iter48.SeekLT("dhfwjzzc@11", "")
iter48.Last()
db.Delete("dhbgmwir@6")
iter48.SeekLT("lhjdjxzgmxk@1", "")
db.SingleDelete("pvuw@8", true /* maybeReplaceDelete */)
iter48.Next("")
db.Delete("znahrxe@12")
db.Delete("xbolxizngr@5")
db.Merge("pvuw@4", "sdffsuwlf")
iter49.Next("")
iter48.SeekGE("hzyxdvgrlc@3", "")
db.DeleteRange("jsbuqwpv@4", "vnck@5")
iter49.SeekPrefixGE("lfmjmlpuzf@14")
iter49.SeekPrefixGE("vyhfcyjvn@2")
db.Get("osraunwzo@11")
iter49.SetBounds("elro@4", "mcsv@4")
iter49.SeekGE("elro@4", "")
iter49.Prev("")
iter49.Next("")
iter48.Next("")
db.DeleteRange("ehcopii@9", "ijvgb@12")
iter48.SeekLT("qvyxa@12", "oeaa@5")
db.Get("rpzqul@11")
db.Get("ulgvuyyuwd@1")
iter48.Close()
iter49.Close()
db.Close()

```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -mod=vendor -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1660669056419909093 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6120746?buildTab=log) with [artifacts](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6120746?buildTab=artifacts#meta) on refs/heads/master @ [c04897ea894f](https://github.com/cockroachdb/pebble/commits/c04897ea894f):


```
        db.SingleDelete("lypv@5", false /* maybeReplaceDelete */) // <nil> #1074
        snap4.Get("dfjuvioqqw@4") // [""] pebble: not found #1075
        iter10.SeekGE("eammeil", "") // [false] <nil> #1076
        iter9.SeekPrefixGE("cleebqesaa@1") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #1077
        snap3.Get("dfjuvioqqw@4") // [""] pebble: not found #1078
        iter10.SeekGE("wiqyj@4", "") // [false] <nil> #1079
        iter10.Prev("") // [true,"drwb","nnsdkokea",<no range>] <nil> #1080
        iter10.SetBounds("ogjji@2", "ufyp") // <nil> #1081
        iter10.SeekGE("ogjji@2", "") // [false] <nil> #1082
        iter10.Prev("") // [false] <nil> #1083
        iter10.Next("") // [false] <nil> #1084
        db.Set("tklhvxyq@5", "haq") // <nil> #1085
        db.SingleDelete("tklhvxyq@5", false /* maybeReplaceDelete */) // <nil> #1086
        iter10.SeekLT("cldj@2", "") // [false] <nil> #1087
        iter9.Next("") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #1088
        iter10.SeekLT("tklhvxyq@2", "") // [false] <nil> #1089
        iter10.SeekLT("wqbhdaps@5", "") // [false] <nil> #1090
        iter9.Next("") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #1091
        db.DeleteRange("lvwunibsz@4", "ufyp") // <nil> #1092
        db.Merge("xryvhpsw@3", "xxepbf") // <nil> #1093
        iter9.SetBounds("wqbhdaps@5", "xmbod@4") // <nil> #1094
        iter9.SeekLT("xmbod@4", "") // [false] <nil> #1095
        iter9.Next("") // [false] <nil> #1096
        iter10.Last() // [false] <nil> #1097
        snap4.Close() // <nil> #1098
        db.Delete("cleebqesaa@1") // <nil> #1099
        iter10.SeekGE("drwb@2", "") // [false] <nil> #1100
        iter10.Next("") // [false] <nil> #1101
        db.Get("wirud") // [""] pebble: not found #1102
        iter10.Next("") // [false] <nil> #1103
        iter9.Next("") // [false] <nil> #1104
        iter9.First() // [false] <nil> #1105
        iter9.Next("") // [false] <nil> #1106
        iter10.Prev("") // [false] <nil> #1107
        iter9.Next("") // [false] <nil> #1108
        db.Set("reinlrjt@5", "d") // <nil> #1109
        iter9.First() // [false] <nil> #1110
        iter10.Prev("") // [false] <nil> #1111
        iter9.SeekGE("ogjji", "") // [false] <nil> #1112
        iter9.Next("") // [false] <nil> #1113
        db.Get("cleebqesaa@1") // [""] pebble: not found #1114
        iter9.SetBounds("lvwunibsz@4", "reinlrjt@5") // <nil> #1115
        iter9.SeekLT("reinlrjt@5", "") // [true,"mbigyr","",<no range>] <nil> #1116
        iter9.Next("") // [false] <nil> #1117
        iter9.Next("") // [false] <nil> #1118
        iter10.Prev("") // [false] <nil> #1119
        iter9.SeekPrefixGE("wajlebtkaa@2") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #1120
        db.Delete("byihpnovyv@1") // <nil> #1121
        iter12 = snap3.NewIter("", "", 2 /* key types */, 0, 0, "@3" /* masking suffix */) // <nil> #1122
    --- FAIL: TestMeta/standard-015 (0.09s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -mod=vendor -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1660689369693919796 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/6123737?buildTab=log) with [artifacts](https://buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/6123737?buildTab=artifacts#meta) on refs/heads/master @ [358c749ef8fa](https://github.com/cockroachdb/pebble/commits/358c749ef8fa):


```
        iter35.Prev("") // [false] <nil> #4124
        db.Ingest(batch29) // <nil> #4125
        iter39.SeekPrefixGE("nmduye") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #4126
        iter38.SeekLT("ofdomv@2", "") // [false] <nil> #4127
        iter40.SeekGE("ogvllst@13", "") // [true,"pgkbvfdxde",<no point>,["pgkbvfdxde","rkufuiwsb")=>{"@2"="dscdgzf"}*] <nil> #4128
        iter39.First() // [true,"tcavhgvy",<no point>,["tcavhgvy","utfjmcha")=>{""="ha","@14"="hfseimwaqirjxn"}*] <nil> #4129
        iter39.SeekPrefixGE("lduftye@4") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #4130
        db.Set("mmfpzpxq@14", "lmft") // <nil> #4131
        iter40.Prev("") // [true,"ogvllst",<no point>,["ogvllst","pgkbvfdxde")=>{"@12"="lsxdrnfjhzspbgky","@2"="dscdgzf"}*] <nil> #4132
        iter42 = db.NewIter("", "", 2 /* key types */, 2, 3, "" /* masking suffix */) // <nil> #4133
        iter35.Prev("") // [false] <nil> #4134
        db.DeleteRange("iktmwm@12", "nefefc@8") // <nil> #4135
        iter40.SetBounds("vsivu@4", "wipydqrdq@4") // <nil> #4136
        iter40.Next("") // [false] <nil> #4137
        db.Delete("yuxmldtkd@2") // <nil> #4138
        db.SingleDelete("dkrwk@1", false /* maybeReplaceDelete */) // <nil> #4139
        iter40.Prev("duptmxbdt@5") // [invalid] <nil> #4140
        db.Get("lsefygpayor@5") // [""] pebble: not found #4141
        iter41.Prev("") // [false] <nil> #4142
        iter42.SetBounds("yuxmldtkd@2", "zpbk@11") // <nil> #4143
        iter42.SeekLT("zpbk@11", "") // [false] <nil> #4144
        iter42.Next("") // [false] <nil> #4145
        iter42.First() // [false] <nil> #4146
        db.Delete("lscziyondda@6") // <nil> #4147
        iter42.Last() // [false] <nil> #4148
        iter40.Next("") // [false] <nil> #4149
        iter42.SeekGE("rdxcwfotx", "") // [false] <nil> #4150
        iter39.SeekGE("hjzgcnln@12", "") // [true,"tcavhgvy",<no point>,["tcavhgvy","utfjmcha")=>{""="ha","@14"="hfseimwaqirjxn"}*] <nil> #4151
        db.Merge("lzypriugt@14", "jjxskpmegfb") // <nil> #4152
        iter39.SeekLT("sbmpa@12", "lscziyondda@6") // [invalid] <nil> #4153
        iter38.First() // [false] <nil> #4154
        db.Delete("roklq@7") // <nil> #4155
        iter41.SeekGE("sovc@7", "") // [true,"tcavhgvy",<no point>,["tcavhgvy","utfjmcha")=>{""="ha","@14"="hfseimwaqirjxn"}*] <nil> #4156
        iter38.Prev("aoxtyrfim@14") // [invalid] <nil> #4157
        db.Delete("pgkbvfdxde@4") // <nil> #4158
        db.Get("xzxamm@1") // [""] pebble: not found #4159
        iter38.Last() // [false] <nil> #4160
        iter39.First() // [true,"tcavhgvy",<no point>,["tcavhgvy","utfjmcha")=>{""="ha","@14"="hfseimwaqirjxn"}*] <nil> #4161
        iter40.Last() // [false] <nil> #4162
        iter40.Last() // [false] <nil> #4163
        db.Get("izvg@4") // [""] pebble: not found #4164
        iter38.SeekGE("brqyynarba@13", "") // [false] <nil> #4165
        iter40.Close() // <nil> #4166
        iter42.Close() // <nil> #4167
        iter41.Close() // <nil> #4168
        iter39.Close() // <nil> #4169
        iter35.Close() // <nil> #4170
        iter38.Close() // <nil> #4171
        // INFO: [JOB 43] WAL created 000188
    --- FAIL: TestMeta/random-012 (2.64s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -mod=vendor -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1660709468554313926 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6123736?buildTab=log) with [artifacts](https://buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/6123736?buildTab=artifacts#meta) on refs/heads/master @ [358c749ef8fa](https://github.com/cockroachdb/pebble/commits/358c749ef8fa):


```
        iter20.SeekLT("kzjngtllp@4", "") // [true,"hpmnyyw",<no point>,["hpmnyyw","smjesqkt")=>{"@8"="zxtyfiwiefvbj"}] <nil> #1953
        iter19.SetBounds("uypdmhvrg", "uypdmhvrg") // <nil> #1954
        iter19.SeekLT("uypdmhvrg", "") // [false] <nil> #1955
        iter20.SeekLT("wtoak@6", "nqpdvywakyu@3") // [valid,"vvbknptdo@7","azdwypetgmlquqmh",<no range>] <nil> #1956
        db.Delete("wtoak@1") // <nil> #1957
        db.Get("uypdmhvrg@2") // [""] pebble: not found #1958
        iter18.Next("nqpdvywakyu@1") // [invalid] <nil> #1959
        iter18.SeekPrefixGE("wtoak@1") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #1960
        iter14.Last() // [true,"yafcpfrbu",<no point>,["yafcpfrbu","yekao@7")=>{"@4"="dklovd"}] <nil> #1961
        db.Set("ipyasg@8", "jngotnqdwvzyf") // <nil> #1962
        db.RangeKeyUnset("hpmnyyw", "ztjnzdwkchp", "@6") // <nil> #1963
        db.SingleDelete("ipyasg@8", false /* maybeReplaceDelete */) // <nil> #1964
        iter20.Prev("") // [true,"sblvcdxfywb@6","iuuvvbehisx",["hpmnyyw","smjesqkt")=>{"@8"="zxtyfiwiefvbj"}] <nil> #1965
        iter14.Next("") // [false] <nil> #1966
        iter20.Close() // <nil> #1967
        iter18.SeekLT("sgjffsczk@8", "") // [false] <nil> #1968
        iter14.Prev("") // [true,"yafcpfrbu",<no point>,["yafcpfrbu","yekao@7")=>{"@4"="dklovd"}*] <nil> #1969
        iter19.SeekGE("spton@4", "") // [false] <nil> #1970
        db.Merge("jfxjpzx@2", "") // <nil> #1971
        iter19.Next("") // [false] <nil> #1972
        db.Delete("blduhsfh@7") // <nil> #1973
        db.Merge("nqpdvywakyu@1", "tilco") // <nil> #1974
        iter19.Prev("") // [false] <nil> #1975
        iter18.Next("") // [false] <nil> #1976
        iter18.First() // [false] <nil> #1977
        db.Delete("hmgce@1") // <nil> #1978
        iter18.SeekGE("yqnn@6", "") // [false] <nil> #1979
        iter14.SetBounds("cfne@4", "hpmnyyw") // <nil> #1980
        iter14.SeekGE("cfne@4", "") // [true,"dfxhlkca",<no point>,["dfxhlkca","enen")=>{"@5"="tdxuituwfptlh"}*] <nil> #1981
        iter14.Prev("") // [false] <nil> #1982
        iter14.Next("") // [true,"dfxhlkca",<no point>,["dfxhlkca","enen")=>{"@5"="tdxuituwfptlh"}*] <nil> #1983
        iter19.SeekGE("irkdwt@5", "") // [false] <nil> #1984
        db.SingleDelete("nlyvjuvve@8", false /* maybeReplaceDelete */) // <nil> #1985
        iter14.Next("") // [false] <nil> #1986
        db.Get("dkdzqi@2") // [""] pebble: not found #1987
        iter14.SeekPrefixGE("dfxhlkca@4") // [false] <nil> #1988
        db.RangeKeySet("ejfdoywsdr", "ieswlgjyxpm", "", "xmcf") // <nil> #1989
        iter19.Next("") // [false] <nil> #1990
        db.Set("qbyd@5", "opzlrafmni") // <nil> #1991
        iter19.SeekGE("cfne@4", "") // [false] <nil> #1992
        iter18.SeekGE("wtoak@6", "") // [false] <nil> #1993
        iter14.SeekLT("spton@2", "") // [true,"dfxhlkca",<no point>,["dfxhlkca","enen")=>{"@5"="tdxuituwfptlh"}] <nil> #1994
        iter18.Next("") // [false] <nil> #1995
        db.Merge("pishvanpkse@5", "wybdevfdkn") // <nil> #1996
        iter18.Close() // <nil> #1997
        iter14.Close() // <nil> #1998
        iter19.Close() // <nil> #1999
        snap13.Close() // <nil> #2000
        // INFO: [JOB 19] WAL created 000226
    --- FAIL: TestMeta/random-008 (0.20s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -mod=vendor -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1660710916678385533 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/data_test.go
+++ b/data_test.go
@@ -227,7 +227,7 @@ func runIterCmd(d *datadriven.TestData, iter *Iterator, closeIter bool) string {
 func parseIterOptions(
 	opts *IterOptions, ref *IterOptions, parts []string,
 ) (foundAny bool, err error) {
-	const usageString = "[lower=<lower>] [upper=<upper>] [key-types=point|range|both] [mask-suffix=<suffix>] [only-durable=<bool>] [table-filter=reuse|none] [point-filters=reuse|none]\n"
+	const usageString = "[lower=<lower>] [upper=<upper>] [key-types=point|range|both] [mask-suffix=<suffix>] [mask-filter=<bool>] [only-durable=<bool>] [table-filter=reuse|none] [point-filters=reuse|none]\n"
 	for _, part := range parts {
 		arg := strings.SplitN(part, "=", 2)
 		if len(arg) != 2 {
@@ -261,7 +261,9 @@ func parseIterOptions(
 		case "mask-suffix":
 			opts.RangeKeyMasking.Suffix = []byte(arg[1])
 		case "mask-filter":
-			opts.RangeKeyMasking.Filter = blockprop.NewMaskingFilter()
+			opts.RangeKeyMasking.Filter = func() BlockPropertyFilterMask {
+				return blockprop.NewMaskingFilter()
+			}
 		case "table-filter":
 			switch arg[1] {
 			case "reuse":

--- a/internal/metamorphic/ops.go
+++ b/internal/metamorphic/ops.go
@@ -577,7 +577,9 @@ func (o *newIterOp) run(t *test, h *history) {
 		},
 	}
 	if opts.RangeKeyMasking.Suffix != nil {
-		opts.RangeKeyMasking.Filter = blockprop.NewMaskingFilter()
+		opts.RangeKeyMasking.Filter = func() pebble.BlockPropertyFilterMask {
+			return blockprop.NewMaskingFilter()
+		}
 	}
 	if o.filterMax > 0 {
 		opts.PointKeyFilters = []pebble.BlockPropertyFilter{
@@ -685,7 +687,9 @@ func (o *iterSetOptionsOp) run(t *test, h *history) {
 		},
 	}
 	if opts.RangeKeyMasking.Suffix != nil {
-		opts.RangeKeyMasking.Filter = blockprop.NewMaskingFilter()
+		opts.RangeKeyMasking.Filter = func() pebble.BlockPropertyFilterMask {
+			return blockprop.NewMaskingFilter()
+		}
 	}
 	if o.filterMax > 0 {
 		opts.PointKeyFilters = []pebble.BlockPropertyFilter{

--- a/iterator_test.go
+++ b/iterator_test.go
@@ -2269,7 +2269,6 @@ func BenchmarkIterator_RangeKeyMasking(b *testing.B) {
 	// Set ignore syncs to true so that each subbenchmark may mutate state and
 	// then revert back to the original state.
 	mem.SetIgnoreSyncs(true)
-	maskingFilter := blockprop.NewMaskingFilter()
 
 	// TODO(jackson): Benchmark lazy-combined iteration versus not.
 	// TODO(jackson): Benchmark seeks.
@@ -2289,7 +2288,9 @@ func BenchmarkIterator_RangeKeyMasking(b *testing.B) {
 				KeyTypes: IterKeyTypePointsAndRanges,
 				RangeKeyMasking: RangeKeyMasking{
 					Suffix: []byte("@100"),
-					Filter: maskingFilter,
+					Filter: func() BlockPropertyFilterMask {
+						return blockprop.NewMaskingFilter()
+					},
 				},
 			}
 			b.Run("forward", func(b *testing.B) {

--- a/options.go
+++ b/options.go
@@ -232,15 +232,17 @@ type RangeKeyMasking struct {
 	Suffix []byte
 	// Filter is an optional field that may be used to improve performance of
 	// range-key masking through a block-property filter defined over key
-	// suffixes. Filter allows Pebble to skip whole point-key blocks containing
-	// point keys with suffixes greater than a covering range-key's suffix.
+	// suffixes. If non-nil, Filter is called by Pebble to construct a
+	// block-property filter mask at iterator creation. The filter is used to
+	// skip whole point-key blocks containing point keys with suffixes greater
+	// than a covering range-key's suffix.
 	//
 	// To use this functionality, the caller must create and configure (through
 	// Options.BlockPropertyCollectors) a block-property collector that records
 	// the maxmimum suffix contained within a block. The caller then must write
 	// and provide a BlockPropertyFilterMask implementation on that same
 	// property. See the BlockPropertyFilterMask type for more information.
-	Filter BlockPropertyFilterMask
+	Filter func() BlockPropertyFilterMask
 }
 
 // BlockPropertyFilterMask extends the BlockPropertyFilter interface for use

--- a/range_keys.go
+++ b/range_keys.go
@@ -183,8 +183,9 @@ func (i *Iterator) constructRangeKeyIter() {
 // result is ignored, and the block is read.
 
 type rangeKeyMasking struct {
-	cmp   base.Compare
-	split base.Split
+	cmp    base.Compare
+	split  base.Split
+	filter BlockPropertyFilterMask
 	// maskActiveSuffix holds the suffix of a range key currently acting as a
 	// mask, hiding point keys with suffixes greater than it. maskActiveSuffix
 	// is only ever non-nil if IterOptions.RangeKeyMasking.Suffix is non-nil.
@@ -203,6 +204,9 @@ type rangeKeyMasking struct {
 func (m *rangeKeyMasking) init(parent *Iterator, cmp base.Compare, split base.Split) {
 	m.cmp = cmp
 	m.split = split
+	if parent.opts.RangeKeyMasking.Filter != nil {
+		m.filter = parent.opts.RangeKeyMasking.Filter()
+	}
 	m.parent = parent
 }
 
@@ -238,7 +242,7 @@ func (m *rangeKeyMasking) SpanChanged(s *keyspan.Span) {
 	if m.maskSpan != nil && m.parent.opts.RangeKeyMasking.Filter != nil {
 		// Update the  block-property filter to filter point keys with suffixes
 		// greater than m.maskActiveSuffix.
-		err := m.parent.opts.RangeKeyMasking.Filter.SetSuffix(m.maskActiveSuffix)
+		err := m.filter.SetSuffix(m.maskActiveSuffix)
 		if err != nil {
 			m.parent.err = err
 		}
@@ -339,7 +343,7 @@ var _ sstable.BoundLimitedBlockPropertyFilter = (*rangeKeyMasking)(nil)
 // Name implements the limitedBlockPropertyFilter interface defined in the
 // sstable package by passing through to the user-defined block property filter.
 func (m *rangeKeyMasking) Name() string {
-	return m.parent.opts.RangeKeyMasking.Filter.Name()
+	return m.filter.Name()
 }
 
 // Intersects implements the limitedBlockPropertyFilter interface defined in the
@@ -351,7 +355,7 @@ func (m *rangeKeyMasking) Intersects(prop []byte) (bool, error) {
 		// No span is actively masking.
 		return true, nil
 	}
-	return m.parent.opts.RangeKeyMasking.Filter.Intersects(prop)
+	return m.filter.Intersects(prop)
 }
 
 // KeyIsWithinLowerBound implements the limitedBlockPropertyFilter interface

--- a/range_keys_test.go
+++ b/range_keys_test.go
@@ -163,7 +163,9 @@ func TestRangeKeys(t *testing.T) {
 				case "mask-suffix":
 					o.RangeKeyMasking.Suffix = []byte(arg.Vals[0])
 				case "mask-filter":
-					o.RangeKeyMasking.Filter = blockprop.NewMaskingFilter()
+					o.RangeKeyMasking.Filter = func() BlockPropertyFilterMask {
+						return blockprop.NewMaskingFilter()
+					}
 				case "lower":
 					o.LowerBound = []byte(arg.Vals[0])
 				case "upper":

--- a/testdata/indexed_batch_mutation
+++ b/testdata/indexed_batch_mutation
@@ -576,3 +576,53 @@ seek-prefix-ge d@2
 d@2: (., [d-"d\x00") @1=foo)
 .
 d@2: (., [d-"d\x00") @2=foo, @1=foo UPDATED)
+
+# Test cloning an iterator with a range-key mask block property filter
+# configured. If the cloned and the clonee iterators have different suffixes
+# configured, their suffixes should be respected. Previously, the
+# RangeKeyMasking.Filter option was a footgun, because it was a single mutable
+# instance. Cloning the iterator without supplying new iterator options would
+# result in two iterators using the same filter.
+
+reset
+----
+
+batch
+range-key-set a e @5 foo
+set b@4 b@4
+----
+
+new-db-iter iter-a
+----
+
+iter iter=iter-a
+set-options mask-suffix=@3 mask-filter=true
+seek-ge a
+next
+next
+----
+.
+a: (., [a-e) @5=foo UPDATED)
+b@4: (b@4, [a-e) @5=foo)
+.
+
+clone from=iter-a to=iter-b refresh-batch=false
+----
+
+iter iter=iter-b
+set-options mask-suffix=@6
+seek-ge a
+next
+----
+.
+a: (., [a-e) @5=foo UPDATED)
+.
+
+iter iter=iter-a
+seek-ge a
+next
+next
+----
+a: (., [a-e) @5=foo UPDATED)
+b@4: (b@4, [a-e) @5=foo)
+.

